src/control/server/mgmt_svc.go

```diff
@@ -96,3 +133,136 @@ func (svc *mgmtSvc) checkReplicaRequest(req proto.Message) error {
 	}
 	return svc.checkSystemRequest(req)
 }
+
+// startBatchLoops kicks off the asynchronous batch processing loops.
+func (svc *mgmtSvc) startBatchLoops(ctx context.Context) {
+	go svc.joinLoop(ctx)
+	go svc.batchLoop(ctx)
+}
+
+// submitBatchRequest submits a message for batch processing and waits for a response.
+func (svc *mgmtSvc) submitBatchRequest(ctx context.Context, msg proto.Message) (proto.Message, error) {
+	respCh := make(batchRespChan, 1)
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case svc.batchReqs <- &batchRequest{msg: msg, ctx: ctx, respCh: respCh}:
+	}
+
+	select {
+	case <-ctx.Done():
+		return nil, ctx.Err()
+	case resp := <-respCh:
+		return resp.msg, resp.err
+	}
+}
+
+// processBatchPoolEvictions processes a batch of PoolEvictReq messages,
+// consolidating them by pool ID in order to minimize the number of
+// dRPC calls to the engine.
+func (svc *mgmtSvc) processBatchPoolEvictions(ctx context.Context, bprChan batchProcessRespChan, reqs []*batchRequest) []*batchRequest {
+	poolReqs := make(map[string][]*batchRequest)
+
+	for _, req := range reqs {
+		msg, ok := req.msg.(*mgmtpb.PoolEvictReq)
+		if !ok {
+			svc.log.Errorf("unexpected message type %T", req.msg)
+			continue
+		}
+		poolReqs[msg.Id] = append(poolReqs[msg.Id], req)
+	}
+
+	for poolID, reqs := range poolReqs {
+		if len(reqs) == 0 {
```

Comment:
this can't happen, as key won't exist unless a valid request has been inserted into the slice

Reply:
Yeah, would be better to put this up at the top of the method instead. I think I added it here before refactoring to add the coalescing bit.